File: leetcode/94. Binary Tree Inorder Traversal/Python3/solution.py
# 94. Binary Tree Inorder Traversal
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        res = []
        stack = []
        while True:
            if root:
                stack.append(root)
                root = root.left
            else:
                if not stack:
                    break
                root = stack.pop()
                res.append(root.val)
                root = root.right
        return res

File: leetcode/94. Binary Tree Inorder Traversal/solution.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        if (root == NULL) return res;

        stack<TreeNode*> s;

        while (1) {
            if (root != NULL) {
                s.push(root);
                root = root->left;
            } else {
                if (s.empty()) break;
                root = s.top();
                res.push_back(root->val);
                s.pop();
                root = root->right;
            }
        }

        return res;
    }
};